### ISSUE TITLE
[2.0.x] Flaky tests:  Cannot find compilation library location for package 'System.Diagnostics.DiagnosticSource'

### ISSUE THREAD

**Issue by natemcmaster** [aspnet/Razor#1721]
I've been seeing Razor tests failing fairly frequently. It doesn't repo on CI, so I'm not sure why it's happening, but on local builds it's pretty consistent. It usually fails about 37 of 156 tests, all with the same error.

Repro steps:
```
git clone --branch rel/2.0.3 --recursive https://github.com/aspnet/Universe.git
build.cmd /p:Configuration=Release
```

```
   Failed   Basic_Runtime
  Error Message:
   System.InvalidOperationException : Cannot find compilation library location for package 'System.Diagnostics.DiagnosticSource'
  Stack Trace:
     at Microsoft.Extensions.DependencyModel.CompilationLibrary.ResolveReferencePaths(ICompilationAssemblyResolver resolver, List`1 assemblies)
     at Microsoft.Extensions.DependencyModel.CompilationLibrary.ResolveReferencePaths()
     at Microsoft.AspNetCore.Mvc.Razor.Extensions.IntegrationTests.CodeGenerationIntegrationTest.<>c.<CreateMvcShimReferences>b__49_0(CompilationLibrary l) in C:\dev\aspnet\Universe\modules\Razor\test\Microsoft.AspNetCo
re.Mvc.Razor.Extensions.Test\IntegrationTests\CodeGenerationIntegrationTest.cs:line 710
     at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
     at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
     at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
     at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
     at Microsoft.AspNetCore.Mvc.Razor.Extensions.IntegrationTests.CodeGenerationIntegrationTest.CreateMvcShimReferences(String mvcShimName) in C:\dev\aspnet\Universe\modules\Razor\test\Microsoft.AspNetCore.Mvc.Razor.Ex
tensions.Test\IntegrationTests\CodeGenerationIntegrationTest.cs:line 712
     at Microsoft.AspNetCore.Mvc.Razor.Extensions.IntegrationTests.CodeGenerationIntegrationTest.Basic_Runtime() in C:\dev\aspnet\Universe\modules\Razor\test\Microsoft.AspNetCore.Mvc.Razor.Extensions.Test\IntegrationTes
ts\CodeGenerationIntegrationTest.cs:line 105
```
cc @Eilon

**Comment by natemcmaster** [comment]
cc @eerhardt have there been any changes in Microsoft.Extensions.DependencyModel since 2.0.0? This is happening in the latest 2.0.3 builds

**Comment by eerhardt** [comment]
There was one change since `2.0.0` - https://github.com/dotnet/core-setup/pull/3134/commits/62b47d66f85c0edbd9009535814c30779724da23 which allowed resolving direct references to be resolved.  The change was in this area, so it is quite possible, but I'm not seeing it.  I'm building 'Universe' locally now to see if I can debug it.

Here's the full history in that branch: https://github.com/dotnet/core-setup/commits/release/2.0.0/src/managed/Microsoft.Extensions.DependencyModel.

**Comment by eerhardt** [comment]
I couldn't get the repo to build using your steps above.  I get:

```
Razor                 | F:\git\Universe\modules\Razor\src\Microsoft.CodeAnalysis.Razor.Workspaces\Microsoft.CodeAnalysis.Razor.Workspaces.csproj : error NU1202: Package Microsoft.CodeAnalysis.Workspaces.Common 2.3.1 is not compatible with net46 (.NETFramework,Version=v4.6). Package Microsoft.CodeAnalysis.Workspaces.Common 2.3.1 supports:
  Razor                 | F:\git\Universe\modules\Razor\src\Microsoft.CodeAnalysis.Razor.Workspaces\Microsoft.CodeAnalysis.Razor.Workspaces.csproj : error NU1202:   - net46 (.NETFramework,Version=v4.6)
  Razor                 | F:\git\Universe\modules\Razor\src\Microsoft.CodeAnalysis.Razor.Workspaces\Microsoft.CodeAnalysis.Razor.Workspaces.csproj : error NU1202:   - netstandard1.3 (.NETStandard,Version=v1.3)
  Razor                 | F:\git\Universe\modules\Razor\src\Microsoft.CodeAnalysis.Razor.Workspaces\Microsoft.CodeAnalysis.Razor.Workspaces.csproj : error NU1202: Package Microsoft.CodeAnalysis.Workspaces.Common 2.3.1 is not compatible with netstandard2.0 (.NETStandard,Version=v2.0). Package Microsoft.CodeAnalysis.Workspaces.Common 2.3.1 supports:
  Razor                 | F:\git\Universe\modules\Razor\src\Microsoft.CodeAnalysis.Razor.Workspaces\Microsoft.CodeAnalysis.Razor.Workspaces.csproj : error NU1202:   - net46 (.NETFramework,Version=v4.6)
  Razor                 | F:\git\Universe\modules\Razor\src\Microsoft.CodeAnalysis.Razor.Workspaces\Microsoft.CodeAnalysis.Razor.Workspaces.csproj : error NU1202:   - netstandard1.3 (.NETStandard,Version=v1.3)
```

which seems really odd.

**Comment by natemcmaster** [comment]
Whoa, that one is weird. Not sure if it's related.

I have a repro on my machine. Not sure yet what the issue is, but here are some clues:
 - it only happens when running tests from .NET Framework. .NET Core appears ok. 
 - System.Diagnostics.DiagnosticsSource/4.4.1 only exists in %USERPROFILE%\.dotnet\x64\sdk\NuGetFallbackFolder. If I copy this over to %USERPROFILE%\.nuget\packages, the tests pass.
 - The code that's failing: https://github.com/aspnet/Razor/blob/e45618eb1b0d5b37091dc38f169c24b4f546cdaf/test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/IntegrationTests/CodeGenerationIntegrationTest.cs#L706-L710
 - What it's trying to do: this test project references a project and manually copies its deps.json file into the OutputPath. We then try to use DependencyContext to get a list of its compile libraries. https://github.com/aspnet/Razor/blob/e45618eb1b0d5b37091dc38f169c24b4f546cdaf/test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test.csproj#L34-L50
 - In my build output, here is the deps.json files for both the test and the referenced assembly, along with the runtimeconfig.dev.json file for the test. The NuGetFallbackFolder is in the list of additionalprobingpaths
```js
// Microsoft.AspNetCore.Mvc.Razor.Extensions.Test.runtimeconfig.dev.json
{
  "runtimeOptions": {
    "additionalProbingPaths": [
      "C:\\Users\\namc\\.dotnet\\store\\|arch|\\|tfm|",
      "C:\\Users\\namc\\.nuget\\packages",
      "C:\\Users\\namc\\.dotnet\\x64\\sdk\\NuGetFallbackFolder"
    ]
  }
}
```

**Comment by rynowak** [comment]
> What it's trying to do: this test project references a project and manually copies its deps.json file into the OutputPath. We then try to use DependencyContext to get a list of its compile libraries.

BTW this is not an obscure or strange scenario, this is something literally every integration test for an MVC site needs to do

**Comment by natemcmaster** [comment]
Found the issue. This isn't a regression in the 2.0.x runtime. It's an issue that @NTaylorMullen  already fixed in dev with this commit https://github.com/aspnet/Razor/commit/127784e0589717f9a302b91ad8a5994592cf54b7. I'm going to cherry-pick back to rel/2.0.x

**Comment by natemcmaster** [comment]
It's just and issue in our test .csproj file. We need a PackageReference to System.Diagnostic.DiagnosticSource in Microsoft.AspNetCore.Mvc.Razor.Extensions.Test.csproj  so the assembly ends up in test output for .NET Framework.

**Comment by natemcmaster** [comment]
Fixed in b537eb448f0b394f4ae74b4618a15b6c2b6d395c
